pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -50,117 +50,142 @@ type NodeManager struct {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
 
-		glog.V(4).Infof("Finding %d datacenters in vc %s", vsi.cfg.Datacenters, vc)
-		if vsi.cfg.Datacenters == "" {
-			datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.conn)
+			err := vsi.conn.Connect(ctx)
 			if err != nil {
-				return err
+				glog.V(4).Info("Discovering node error vc:", err)
 			}
-		} else {
-			datacenters := strings.Split(vsi.cfg.Datacenters, ",")
-			for _, dc := range datacenters {
-				dc = strings.TrimSpace(dc)
-				if dc == "" {
-					continue
-				}
-				datacenterObj, err := vclib.GetDatacenter(ctx, vsi.conn, dc)
+
+			if vsi.cfg.Datacenters == "" {
+				datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.conn)
 				if err != nil {
-					return err
+					glog.V(4).Info("Discovering node error dc:", err)
+				}
+			} else {
+				datacenters := strings.Split(vsi.cfg.Datacenters, ",")
+				for _, dc := range datacenters {
+					dc = strings.TrimSpace(dc)
+					if dc == "" {
+						continue
+					}
+					datacenterObj, err := vclib.GetDatacenter(ctx, vsi.conn, dc)
+					if err != nil {
+						glog.V(4).Info("Discovering node error dc:", err)
+					}
+					datacenterObjs = append(datacenterObjs, datacenterObj)
 				}
-				datacenterObjs = append(datacenterObjs, datacenterObj)
 			}
-		}
-		glog.V(4).Infof("Found %d datacenters in vc %s", len(datacenterObjs), vc)
-
-		type VmSearchResult struct {
-			vm *vclib.VirtualMachine
-			datacenter *vclib.Datacenter
-			err error
-		}
 
-		c1 := make(chan *VmSearchResult)
-		nodeUUID := node.Status.NodeInfo.SystemUUID
-		for _, datacenterObj := range datacenterObjs {
-			glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
-			go func(dc *vclib.Datacenter) {
-				vm, err := dc.GetVMByUUID(ctx, nodeUUID)
-				c1 <- &VmSearchResult{
-					vm:         vm,
-					err:        err,
+			for _, datacenterObj := range datacenterObjs {
+				glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
+				queueChannel <- &VmSearch{
+					vc:         vc,
 					datacenter: datacenterObj,
 				}
-			}(datacenterObj)
-		}
-
-		for i := 0; i < len(datacenterObjs); i++ {
-			res := <-c1
-			if res.vm != nil {
-				glog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
-					node.Name, res.vm, vc, res.datacenter.Name())
-
-				nodeInfo := &NodeInfo{dataCenter: datacenterObjs[i], vm: res.vm, vcServer: vc}
-				nm.nodeInfoLock.Lock()
-				nm.nodeInfoMap[node.ObjectMeta.Name] = nodeInfo
-				nm.nodeInfoLock.Unlock()
-			} else {
-				glog.V(4).Infof("Did not find node %s in vc=%s and datacenter=%s",
-					node.Name, vc, res.datacenter.Name())
 			}
 		}
+		close(queueChannel)
+	}()
+
+	for i := 0; i < POOL_SIZE; i++ {
+		go func() {
+			for res:= range queueChannel {
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				vm, err := res.datacenter.GetVMByUUID(ctx, nodeUUID)
```

Comment:
Can you please log the err if it is not nil?

Reply:
Done